File: src/node/harp-approver.mjs
// harp-approver.mjs
// Mobile Approver — generates MA keys (first run), decrypts artifact, signs decision
// Mirrors Harp.Approver/Program.cs

import { readFileSync, writeFileSync, existsSync, mkdirSync } from 'node:fs';
import { createInterface } from 'node:readline';
import {
    toB64Url, fromB64Url, sha256Hex, deriveKey,
    createX25519Keypair, x25519DeriveShared,
    createEd25519Keypair, ed25519Sign, ed25519Verify,
    xchachaDecrypt, randomBytes,
} from './lib/crypto-helpers.mjs';
import { jcsCanonicalize } from './lib/canonical-json.mjs';

const MA_KEYS_FILE = 'C:\\tmp\\harp\\ma-keys.json';
const ARTIFACT_WIRE_FILE = 'C:\\tmp\\harp\\artifact-wire.json';
const DECISION_FILE = 'C:\\tmp\\harp\\decision.json';

function preview(s) {
    return s.length <= 18 ? s : s.substring(0, 18) + '...';
}

function prompt(question) {
    const rl = createInterface({ input: process.stdin, output: process.stdout });
    return new Promise((resolve) => {
        rl.question(question, (answer) => {
            rl.close();
            resolve(answer);
        });
    });
}

// ──────────────── Key Bootstrap ────────────────

function ensureKeysExist() {
    if (existsSync(MA_KEYS_FILE)) return;

    console.log('Generating MA keys...');

    // X25519 keypair for payload decryption
    const kxKp = createX25519Keypair();

    // Ed25519 keypair for decision signing
    const signKp = createEd25519Keypair();

    const keys = {
        maKxPubRawB64Url: toB64Url(kxKp.publicKey),
        maKxPrivRawB64Url: toB64Url(kxKp.privateKey),
        maEd25519PubRawB64Url: toB64Url(signKp.publicKey),
        maEd25519PrivRawB64Url: toB64Url(signKp.privateKey),
        signerKeyId: 'ma-key-1',
    };

    mkdirSync('C:\\tmp\\harp', { recursive: true });
    writeFileSync(MA_KEYS_FILE, JSON.stringify(keys, null, 2));
    console.log(`✅ Wrote ${MA_KEYS_FILE}`);
}

// ──────────────── Main ────────────────

ensureKeysExist();

const keys = JSON.parse(readFileSync(MA_KEYS_FILE, 'utf8'));
const maKxPrivRaw = fromB64Url(keys.maKxPrivRawB64Url);
const maKxPubB64Url = keys.maKxPubRawB64Url;
const maSignPrivRaw = fromB64Url(keys.maEd25519PrivRawB64Url);
const maSignPubRaw = fromB64Url(keys.maEd25519PubRawB64Url);
const signerKeyId = keys.signerKeyId;

if (!existsSync(ARTIFACT_WIRE_FILE)) {
    console.log(`Missing ${ARTIFACT_WIRE_FILE}. Run harp-executor.mjs first.`);
    process.exit(0); // First run was just key generation
}

const artifactWire = JSON.parse(readFileSync(ARTIFACT_WIRE_FILE, 'utf8'));

// Basic enc sanity
if (artifactWire.enc.encAlg !== 'XChaCha20-Poly1305') {
    console.log(`❌ Unsupported encAlg: ${artifactWire.enc.encAlg}`);
    process.exit(1);
}
if (artifactWire.enc.kdf !== 'X25519+HKDF-SHA256') {
    console.log(`❌ Unsupported kdf: ${artifactWire.enc.kdf}`);
    process.exit(1);
}

// Rebuild AAD (must match HE)
const aadObj = {
    requestId: artifactWire.requestId,
    artifactType: artifactWire.artifactType,
    repoRef: artifactWire.repoRef,
    createdAt: artifactWire.createdAt,
    expiresAt: artifactWire.expiresAt,
    artifactHashAlg: artifactWire.artifactHashAlg,
    artifactHash: artifactWire.artifactHash,
};
const aad = Buffer.from(jcsCanonicalize(aadObj), 'utf8');

// ──── Key agreement + HKDF ────

const heKxPubRaw = fromB64Url(artifactWire.enc.heKxPub);
const sharedSecret = x25519DeriveShared(maKxPrivRaw, heKxPubRaw);

const salt = fromB64Url(artifactWire.enc.salt);
const info = Buffer.from(artifactWire.enc.info || 'HARP-XCHACHA-PAYLOAD-V1', 'utf8');
const keyMaterial = deriveKey(sharedSecret, salt, info, 32);

// ──── AEAD decrypt ────

const nonce = fromB64Url(artifactWire.enc.nonce);
const ciphertext = fromB64Url(artifactWire.enc.ciphertext);
const tag = fromB64Url(artifactWire.enc.tag);

let plaintext;
try {
    plaintext = xchachaDecrypt(keyMaterial, nonce, ciphertext, tag, aad);
} catch (err) {
    console.log(`❌ Decryption/auth failed: ${err.message}`);
    process.exit(1);
}

const payloadJson = plaintext.toString('utf8');
const payloadObj = JSON.parse(payloadJson);

// ──── Verify artifactHash by reconstructing plaintext artifact without artifactHash ────

const artifactWithoutHash = {
    requestId: artifactWire.requestId,
    artifactType: artifactWire.artifactType,
    repoRef: artifactWire.repoRef,
    createdAt: artifactWire.createdAt,
    expiresAt: artifactWire.expiresAt,
    payload: payloadObj,
    artifactHashAlg: artifactWire.artifactHashAlg,
};

const recomputed = sha256Hex(jcsCanonicalize(artifactWithoutHash));
if (recomputed.toLowerCase() !== artifactWire.artifactHash.toLowerCase()) {
    console.log('❌ Hash mismatch. Refuse.');
    console.log(`Expected: ${artifactWire.artifactHash}`);
    console.log(`Actual:   ${recomputed}`);
    process.exit(1);
}

console.log('✅ Payload decrypted and artifactHash verified.');
console.log();
console.log('----- REVIEW PAYLOAD -----');
console.log(payloadJson);
console.log('--------------------------');
console.log();

const answer = (await prompt('Approve? (y/n): ')).trim().toLowerCase();
const decisionValue = answer === 'y' ? 'allow' : 'deny';

const decisionExpires = new Date(Date.now() + 10 * 60 * 1000);
const decisionNonce = toB64Url(randomBytes(16));

// Build DecisionSignable and sign
const signable = {
    requestId: artifactWire.requestId,
    artifactHashAlg: artifactWire.artifactHashAlg,
    artifactHash: artifactWire.artifactHash,
    repoRef: artifactWire.repoRef,
    decision: decisionValue,
    scope: 'once',
    expiresAt: decisionExpires.toISOString(),
    nonce: decisionNonce,
    sigAlg: 'Ed25519',
    signerKeyId,
};

const signableCanon = Buffer.from(jcsCanonicalize(signable), 'utf8');
const signature = ed25519Sign(maSignPrivRaw, signableCanon);

// Optional self-verify
if (!ed25519Verify(maSignPubRaw, signableCanon, signature)) {
    console.log('❌ Signature self-verify failed.');
    process.exit(1);
}

const decision = {
    ...signable,
    signature: toB64Url(signature),
};

writeFileSync(DECISION_FILE, JSON.stringify(decision, null, 2));
console.log(`✅ Wrote ${DECISION_FILE} (${decision.decision})`);

File: src/rust/src/bin/harp_approver.rs
//! Mobile Approver — generates MA keys (first run), decrypts artifact, signs decision.

use chrono::{Duration, Utc};
use serde_json::json;
use std::fs;
use std::io::{self, Write};
use std::process;

use harp_rust::canonical_json;
use harp_rust::crypto_helpers::{
    create_ed25519_keypair, create_x25519_keypair, derive_key, ed25519_sign, ed25519_verify,
    from_b64url, random_bytes, sha256_hex, to_b64url, xchacha_decrypt,
};
use harp_rust::models::{ArtifactWire, Decision, DecisionSignable, MaKeys};

const MA_KEYS_FILE: &str = r"C:\tmp\harp\ma-keys.json";
const ARTIFACT_WIRE_FILE: &str = r"C:\tmp\harp\artifact-wire.json";
const DECISION_FILE: &str = r"C:\tmp\harp\decision.json";

fn ensure_keys_exist() {
    if fs::metadata(MA_KEYS_FILE).is_ok() {
        return;
    }

    println!("Generating MA keys...");

    let kx_kp = create_x25519_keypair();
    let sign_kp = create_ed25519_keypair();

    let keys = MaKeys {
        ma_kx_pub_raw_b64_url: to_b64url(&kx_kp.public_key),
        ma_kx_priv_raw_b64_url: to_b64url(&kx_kp.private_key),
        ma_ed25519_pub_raw_b64_url: to_b64url(&sign_kp.public_key),
        ma_ed25519_priv_raw_b64_url: to_b64url(&sign_kp.private_key),
        signer_key_id: "ma-key-1".into(),
    };

    fs::create_dir_all(r"C:\tmp\harp").ok();
    let out = serde_json::to_string_pretty(&keys).unwrap();
    fs::write(MA_KEYS_FILE, out).unwrap();
    println!("✅ Wrote {}", MA_KEYS_FILE);
}

fn main() {
    ensure_keys_exist();

    let keys: MaKeys = serde_json::from_str(&fs::read_to_string(MA_KEYS_FILE).unwrap()).unwrap();

    let ma_kx_priv_raw = from_b64url(&keys.ma_kx_priv_raw_b64_url).unwrap();
    let ma_sign_priv_raw = from_b64url(&keys.ma_ed25519_priv_raw_b64_url).unwrap();
    let ma_sign_pub_raw = from_b64url(&keys.ma_ed25519_pub_raw_b64_url).unwrap();

    if fs::metadata(ARTIFACT_WIRE_FILE).is_err() {
        println!("Missing {}. Run harp-executor first.", ARTIFACT_WIRE_FILE);
        process::exit(0);
    }

    let artifact_wire: ArtifactWire =
        serde_json::from_str(&fs::read_to_string(ARTIFACT_WIRE_FILE).unwrap()).unwrap();

    if artifact_wire.enc.enc_alg != "XChaCha20-Poly1305" {
        eprintln!("❌ Unsupported encAlg: {}", artifact_wire.enc.enc_alg);
        process::exit(1);
    }
    if artifact_wire.enc.kdf != "X25519+HKDF-SHA256" {
        eprintln!("❌ Unsupported kdf: {}", artifact_wire.enc.kdf);
        process::exit(1);
    }

    // Rebuild AAD
    let aad_obj = json!({
        "requestId": artifact_wire.request_id,
        "artifactType": artifact_wire.artifact_type,
        "repoRef": artifact_wire.repo_ref,
        "createdAt": artifact_wire.created_at,
        "expiresAt": artifact_wire.expires_at,
        "artifactHashAlg": artifact_wire.artifact_hash_alg,
        "artifactHash": artifact_wire.artifact_hash
    });
    let aad = canonical_json::canonicalize(&aad_obj).unwrap();

    // Key agreement + HKDF
    let he_kx_pub_raw = from_b64url(&artifact_wire.enc.he_kx_pub).unwrap();
    let mut my_priv = [0u8; 32];
    my_priv.copy_from_slice(&ma_kx_priv_raw);
    let mut peer_pub = [0u8; 32];
    peer_pub.copy_from_slice(&he_kx_pub_raw);
    let shared_secret = harp_rust::crypto_helpers::x25519_derive_shared(&my_priv, &peer_pub);

    let salt = from_b64url(&artifact_wire.enc.salt).unwrap();
    let info = if artifact_wire.enc.info.is_empty() {
        "HARP-XCHACHA-PAYLOAD-V1"
    } else {
        &artifact_wire.enc.info
    };
    let key_material = derive_key(&shared_secret, &salt, info.as_bytes(), 32);

    // AEAD decrypt
    let nonce = from_b64url(&artifact_wire.enc.nonce).unwrap();
    let ciphertext = from_b64url(&artifact_wire.enc.ciphertext).unwrap();
    let tag = from_b64url(&artifact_wire.enc.tag).unwrap();

    let plaintext = match xchacha_decrypt(&key_material, &nonce, &ciphertext, &tag, aad.as_bytes())
    {
        Ok(pt) => pt,
        Err(e) => {
            eprintln!("❌ Decryption/auth failed: {}", e);
            process::exit(1);
        }
    };

    let payload_json = String::from_utf8(plaintext).unwrap();
    let payload_obj: serde_json::Value = serde_json::from_str(&payload_json).unwrap();

    // Verify artifactHash
    let artifact_without_hash = json!({
        "requestId": artifact_wire.request_id,
        "artifactType": artifact_wire.artifact_type,
        "repoRef": artifact_wire.repo_ref,
        "createdAt": artifact_wire.created_at,
        "expiresAt": artifact_wire.expires_at,
        "payload": payload_obj,
        "artifactHashAlg": artifact_wire.artifact_hash_alg
    });

    let canon_str = canonical_json::canonicalize(&artifact_without_hash).unwrap();
    let recomputed = sha256_hex(&canon_str);

    if recomputed.to_lowercase() != artifact_wire.artifact_hash.to_lowercase() {
        eprintln!("❌ Hash mismatch. Refuse.");
        eprintln!("Expected: {}", artifact_wire.artifact_hash);
        eprintln!("Actual:   {}", recomputed);
        process::exit(1);
    }

    println!("✅ Payload decrypted and artifactHash verified.");
    println!();
    println!("----- REVIEW PAYLOAD -----");
    println!("{}", payload_json);
    println!("--------------------------");
    println!();

    print!("Approve? (y/n): ");
    io::stdout().flush().unwrap();
    let mut answer = String::new();
    io::stdin().read_line(&mut answer).unwrap();
    let decision_value = if answer.trim().to_lowercase() == "y" {
        "allow"
    } else {
        "deny"
    };

    let decision_expires = Utc::now() + Duration::minutes(10);
    let decision_nonce = to_b64url(&random_bytes(16));

    let signable = DecisionSignable {
        request_id: artifact_wire.request_id.clone(),
        artifact_hash_alg: artifact_wire.artifact_hash_alg.clone(),
        artifact_hash: artifact_wire.artifact_hash.clone(),
        repo_ref: artifact_wire.repo_ref.clone(),
        decision: decision_value.into(),
        scope: "once".into(),
        expires_at: decision_expires.to_rfc3339(),
        nonce: decision_nonce,
        sig_alg: "Ed25519".into(),
        signer_key_id: keys.signer_key_id.clone(),
    };

    let signable_canon = canonical_json::canonicalize_bytes(&signable).unwrap();
    let signature = ed25519_sign(&ma_sign_priv_raw, &signable_canon);

    // Self-verify
    if !ed25519_verify(&ma_sign_pub_raw, &signable_canon, &signature) {
        eprintln!("❌ Signature self-verify failed.");
        process::exit(1);
    }

    let decision = Decision {
        request_id: signable.request_id,
        artifact_hash_alg: signable.artifact_hash_alg,
        artifact_hash: signable.artifact_hash,
        repo_ref: signable.repo_ref,
        decision: signable.decision.clone(),
        scope: signable.scope,
        expires_at: signable.expires_at,
        nonce: signable.nonce,
        sig_alg: signable.sig_alg,
        signer_key_id: signable.signer_key_id,
        signature: to_b64url(&signature),
    };

    let out = serde_json::to_string_pretty(&decision).unwrap();
    fs::write(DECISION_FILE, out).unwrap();
    println!("✅ Wrote {} ({})", DECISION_FILE, decision.decision);
}
